SO-4080: optimize bool index queries

cmark opened the pull request:
This PR optimizes bool index queries to reduce (and hopefully eliminate) the chance of hitting search errors:

1. Merge term filters specified on the same field to reduce the number of bool query clauses
2. Flatten deep bool query hierarchies to reduce the chance of getting HTTP 400 error due to `indices.query.bool.max_nested_depth` setting, which defaults to depth of `20`.

cmark committed:
feat(index): optimize bool index queries as much as possible

1. merge term filters specified on the same field to reduce the number
of bool query clauses
2. flatten deep bool query hierarchies to reduce the chance of getting
HTTP 400 error due to `indices.query.bool.max_nested_depth` setting,
which defaults to depth of `20`.

cmark committed:
fix(index): throw IllegalArgumentException when Elasticsearch reports...

...an HTTP 400 error

apeteri commented on commons/com.b2international.index/src/com/b2international/index/es/EsDocumentSearcher.java:
```diff
@@ -183,6 +185,9 @@ public EsDocumentSearcher(EsIndexAdmin admin, ObjectMapper mapper) {
 		try {
 			response = client.search(req);
 		} catch (Exception e) {
+			if (e instanceof ElasticsearchStatusException && ((ElasticsearchStatusException) e).status() == RestStatus.BAD_REQUEST) {
+				throw new IllegalArgumentException(e.getMessage(), e);
+			}
```
Please add this block to a separate `catch (ElasticSearchStatusException e) { ... }` clause above the existing "catch-all" block.

cmark replied:
Leaving as is now, as the current catch-all block handles the logging and generic index exception in case of any other ES exception thrown by the search request.

apeteri replied:
Following discussion with @cmark, non-400 errors should still be logged, so the `catch` block can be left as is.

apeteri commented on commons/com.b2international.index/src/com/b2international/index/query/AbstractExpressionBuilder.java:
```diff
@@ -72,13 +78,90 @@ public B setMinimumNumberShouldMatch(int minShouldMatch) {
 		if (mustClauses.isEmpty() && mustNotClauses.isEmpty() && shouldClauses.isEmpty() && filterClauses.isEmpty()) {
 			return Expressions.matchAll();
-		} else if (mustClauses.isEmpty() && mustNotClauses.isEmpty() && shouldClauses.isEmpty() && filterClauses.size() == 1) {
+		} else if (isSingleFilter()) {
 			// shortcut to reduce number of nested Boolean clauses
 			return filterClauses.get(0);
```
We could support `isSingleShould()` here as well, as you write in the final `else` block below (single "shoulds" _must_ match). 

Also if we want to be really precise, the extracted single filter clause should be wrapped in a non-scoring query wrapper to achieve the same effect as a boolean filter &ndash; which is not contributing to the final document score.

```suggestion
		} else if (isSingleFilter()) {
			// shortcut to reduce number of nested Boolean clauses
			return filterClauses.get(0);
			// return new NoScoreExpression(filterClauses.get(0)); <-- this class does not exist yet
		} else if (isSingleShould()) {
			return filterClauses.get(0);
```

apeteri commented on commons/com.b2international.index/src/com/b2international/index/query/AbstractExpressionBuilder.java:
```diff
@@ -72,13 +78,90 @@ public B setMinimumNumberShouldMatch(int minShouldMatch) {
+			if (expression instanceof BoolExpression) {
+				BoolExpression bool = (BoolExpression) expression;
+				if (bool.isShouldOnly() && bool.minShouldMatch() == 1) {
+					shouldClauses.addAll(bool.filterClauses());
```
```suggestion
					shouldClauses.addAll(bool.shouldClauses());
```

apeteri replied:
(I understand that starting with ES 8.0 we don't have a limit on the number of "should"-clauses. Don't we want to restrain ourselves regardless? 😃 So `addAll` would only be called if the resulting clause list remains under the previous 1024 limit.)

cmark replied:
Not sure, I would rely on the ES to tell us if a query is too big to execute, (has too deep bool clause structure or too many clauses or whatevs). We can monitor the executed queries and add this later on if needed.

apeteri commented on commons/com.b2international.index/src/com/b2international/index/query/AbstractExpressionBuilder.java:
```diff
@@ -72,13 +78,90 @@ public B setMinimumNumberShouldMatch(int minShouldMatch) {
+				}
+				// replace all clauses with a single expression
+				clauses.add(Expressions.matchAnyObject(field, values));
+				clauses.removeAll(termExpressions);
```
Hopefully the `equals` + `hashCode` implementations of `SingleArgumentPredicate` and `SetPredicate` do not interfere with this remove operation. On another note, they could be used to remove duplicate expressions!

cmark replied:
Yeah, I'll swap the remove and add call order to remove the old expressions first then add the new one, it can happen that the new expression is the same as one of the removed ones and in that case, the resulting boolean expression will be empty, which is wrong.

cmark committed:
fix(index): wrap revision queries properly in must/filter clauses...

...instead of always using must (which is mainly for scoring)

cmark committed:
fix(index): remove must/filter clauses from term filter merging

They cause the resulting boolean logic to change from and to or, which
leads to incorrect results.

cmark committed:
fix(index): remove unnecessary query wrapping when building ES query

cmark committed:
fix(index): do not wrap single should clause into bool query

cmark committed:
feat(index): reduce must/filter clauses to single expression...

...by using the intersection of the specified values

cmark committed:
fix(index): generate new Set instead of reusing SetView from...

...Sets.intersection when collapsing term filters

apeteri reviewed (approved):
LGTM!
